File: elixir/pe009.exs
defmodule ProjectEuler do
  def pe009(n) do
    pythagorean_triplet?(n, 1, 2, 3)
  end
   
  defp equal_sum?(a, b, c, n) do
    n == a + b + c
  end

  defp equal_squares?(a, b, c) do
    a * a + b * b == c * c
  end

  defp both_match?(a, b, c, n) do
    equal_sum?(a, b, c, n) and equal_squares?(a, b, c)
  end

  defp pythagorean_triplet?(n, a, b, c) when c < n do
    if both_match?(a, b, c, n) do
      a * b * c
    else
      pythagorean_triplet?(n, a, b, c + 1)
    end
  end

  defp pythagorean_triplet?(n, a, b, c) when b < n do
    if both_match?(a, b, c, n) do
      a * b * c
    else
      pythagorean_triplet?(n, a, b + 1, 1)
    end
  end

  defp pythagorean_triplet?(n, a, b, c) when a < n do
    if both_match?(a, b, c, n) do
      a * b * c
    else
      pythagorean_triplet?(n, a + 1, 1, 1)
    end
  end
end

IO.puts(ProjectEuler.pe009(1000))

File: haskell/pe009.hs
-- this computes projecteuler.net problem 009
equal_sum :: Int -> Int -> Int -> Int -> Bool
equal_sum n a b c = n == a + b + c

equal_squares :: Int -> Int -> Int -> Bool
equal_squares a b c = a * a + b * b == c * c

both_match :: Int -> Int -> Int -> Int -> Bool
both_match n a b c = equal_sum n a b c && equal_squares a b c

pythagoran_triplet :: Int -> Int -> Int -> Int -> Int
pythagoran_triplet n a b c | c < n =
  if both_match n a b c then
    a * b * c
  else
    pythagoran_triplet n a b (c + 1)

pythagoran_triplet n a b c | b < n =
  if both_match n a b c then
    a * b * c
  else
    pythagoran_triplet n a (b + 1) 1

pythagoran_triplet n a b c | a < n =
  if both_match n a b c then
    a * b * c
  else
    pythagoran_triplet n (a + 1) 1 1

pe009 :: Int -> Int
pe009 n = pythagoran_triplet n 1 2 3

main = print(pe009(1000))
